scripts/triggers-end-to-end-tests/cli.ts

```diff
@@ -0,0 +1,69 @@
+import * as subprocess from "child_process";
+
+export class CLIProcess {
+  process?: subprocess.ChildProcess;
+
+  constructor(private readonly name: string, private readonly workdir: string) {}
+
+  start(
+    cmd: string,
+    project: string,
+    additionalArgs: string[],
+    logDoneFn?: (d: unknown) => unknown
+  ): Promise<void> {
+    const args = [cmd, "--project", project];
+
+    if (additionalArgs) {
+      args.push(...additionalArgs);
+    }
+
+    this.process = subprocess.spawn("firebase", args, { cwd: this.workdir });
+    if (!this.process) {
+      throw new Error("Failed to start firebase CLI");
+    }
+
+    this.process.stdout.on("data", (data: unknown) => {
+      process.stdout.write(`[${this.name} stdout] ` + data);
+    });
+
+    this.process.stderr.on("data", (data: unknown) => {
+      console.log(`[${this.name} stderr] ` + data);
+    });
+
+    let started: Promise<void>;
+    if (logDoneFn) {
+      started = new Promise((resolve) => {
+        this.process?.stdout.on("data", (data: unknown) => {
+          if (logDoneFn(data)) {
```

Comment:
Could this hang forever?  Shouldn't we still resolve on `close`?

Reply:
This block of code tricked me the first time. If we have this function, we make this resolving promise. If we don't, we do one on 'close' below.

Basically, logDoneFn looks for a string in the data that comes through and says "started" when that happens (logDoneFn is user-supplied). If that fn isn't provided, it just waits until the process closes.

Reply:
Yes but what if `logDoneFn(data)` never returns `true`?  Then the function will never hit `resolve()`

Reply:
Gotcha. After a quick discussion, we're going to reject if logDoneFn never succeeds :)